dataline-integrations/singer/postgres/source/src/test-integration/java/io/dataline/integrations/io/dataline/integration_tests/sources/SingerPostgresSourceTest.java

```diff
@@ -0,0 +1,237 @@
+import io.dataline.workers.protocol.singer.SingerMessageTracker;
+import io.dataline.workers.singer.SingerCheckConnectionWorker;
+import io.dataline.workers.singer.SingerDiscoverSchemaWorker;
+import io.dataline.workers.singer.SingerTapFactory;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.utility.MountableFile;
+
+public class SingerPostgresSourceTest {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SingerPostgresSourceTest.class);
+  private static final String IMAGE_NAME = "dataline/integration-singer-postgres-source";
+  private static final Path TESTS_PATH = Path.of("/tmp/dataline_integration_tests");
+
+  private PostgreSQLContainer psqlDb;
+  private ProcessBuilderFactory pbf;
+  private Path workspaceRoot;
+  private Path jobRoot;
+
+  @BeforeEach
+  public void init() throws IOException {
+    psqlDb = new PostgreSQLContainer();
+    psqlDb.start();
+
+    PostgreSQLContainerHelper.runSqlScript(MountableFile.forClasspathResource("simple_postgres_init.sql"), psqlDb);
+    Files.createDirectories(TESTS_PATH);
+    workspaceRoot = Files.createTempDirectory(TESTS_PATH, "dataline-integration");
+    jobRoot = workspaceRoot.resolve("job");
+    Files.createDirectories(jobRoot);
+
+    pbf = new DockerProcessBuilderFactory(workspaceRoot, workspaceRoot.toString(), "host");
+  }
+
+  @AfterEach
+  public void cleanUp() {
+    psqlDb.stop();
+  }
+
+  @Test
+  public void testReadFirstTime() throws IOException, InvalidCredentialsException {
+    SingerTapFactory singerTapFactory = new SingerTapFactory(IMAGE_NAME, pbf, new SingerDiscoverSchemaWorker(IMAGE_NAME, pbf));
+
+    Schema schema = Jsons.deserialize(MoreResources.readResource("simple_postgres_source_schema.json"), Schema.class);
+
+    // select all tables and all columns
+    schema.getTables().forEach(t -> t.setSelected(true));
+    schema.getTables().forEach(t -> t.getColumns().forEach(c -> c.setSelected(true)));
+
+    StandardSync syncConfig = new StandardSync().withSyncMode(StandardSync.SyncMode.FULL_REFRESH).withSchema(schema);
+    SourceConnectionImplementation sourceImpl =
+        new SourceConnectionImplementation().withConfiguration(Jsons.jsonNode(getDbConfig()));
+
+    StandardTapConfig tapConfig = new StandardTapConfig()
+        .withStandardSync(syncConfig)
+        .withSourceConnectionImplementation(sourceImpl);
+
+    Stream<SingerMessage> singerMessageStream = singerTapFactory.create(tapConfig, jobRoot);
```

Comment:
I thought for the integration-specific tests we wanted to just test that the config -> output works properly, and not test the tap factory/dataline runner infrastructure?

Reply:
I think once we have comprehensive knowledge of all the things that can come out of a singer tap at compile-time, I agree with that approach. But initially (and maybe for a while), it's going to be hard to know that. Given this difficulty, IMO the better tradeoff to make is to have some duplication of test functionality for the benefit of providing higher confidence that all the machinery works well together, rather than the other way around. 

For example: `tap-postgres` produced a `SCHEMA` message which failed our JsonSchema encoding of singer messages. It also produced a `RECORD` message which also failed our JsonSchema encoding for a different reason. And it exposed that there is a different type of message, `ACTIVATE_VERSION` that is defacto used by many taps, but isn't documented in the spec because it's still experimental. If we didn't test the `TapFactory` in this test and used a processbuilderfactory etc.. we would have only found out those issues when running the full application. 

So I think for now the approach should be for now: port any such findings into unit tests where applicable (e.g: add those funky records types into `SingerTapFactoryTest`) and continue to use wider machinery than is necessary in integration tests until we develop enough confidence that our unit tests are comprehensive enough that we can use more barebones structs in our integration tests (which potentially will come once integrations are fully encapsulated in docker images e.g: when stdout of a docker image outputs dataline data models instead of singer data models that we then convert in the application). 


WDYT?

Reply:
Talked offline. Sounds good.